File: easy/python/symmetric_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


from typing import Optional


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        def helper(left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
            if not left and not right:
                return True
            elif not left and right or left and not right:
                return False
            elif left and right and left.val != right.val:
                return False

            return helper(left.left, right.right) and helper(left.right, right.left)

        return helper(root.left, root.right)

File: easy/ruby/symmetric_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
  helper = lambda do |node1, node2|
    return true if !node1 && !node2
    return false if !node1 || !node2
    return false if node1.val != node2.val

    helper.call(node1.left, node2.right) &&
      helper.call(node1.right, node2.left)
  end

  helper.call(root.left, root.right)
end
